[Flang][OpenMP] Initial defaultmap implementation

agozillon opened the pull request:
This aims to implement most of the initial arguments for defaultmap aside from firstprivate and none, and some of the more recent OpenMP 6 additions which will come in subsequent updates (with the OpenMP 6 variants needing parsing/semantic support first).

agozillon commented:
Small ping on this for reviews if at all possible, thank you very much ahead of time :-)

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2272,6 +2276,129 @@ genTargetOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
   };
 
+  auto getDefaultmapIfPresent = [&](mlir::Type varType) {
+    using defMap = clause::Defaultmap;
```
```suggestion
    using DefMap = clause::Defaultmap;
```

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2272,6 +2276,129 @@ genTargetOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
     return size <= ptrSize && align <= ptrAlign;
   };
 
+  auto getDefaultmapIfPresent = [&](mlir::Type varType) {
```
Nit: This and `getImplicitMapTypeAndKind` below are probably better as `static` functions rather than lambdas, to help with readability.

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2327,62 +2454,32 @@ genTargetOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
-                             baseOp.getLoc());
+      auto mapFlagAndKind = getImplicitMapTypeAndKind(eleType, sym);
+
+      llvm::SmallVector<mlir::Value> bounds =
+          fir::factory::genImplicitBoundsOps<mlir::omp::MapBoundsOp,
+                                             mlir::omp::MapBoundsType>(
+              firOpBuilder, info, dataExv,
+              semantics::IsAssumedSizeArray(sym.GetUltimate()),
+              converter.getCurrentLocation());
```
Nit: Is moving this block an intended change?

skatrak commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2272,6 +2276,129 @@ genTargetOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
 
+  auto getDefaultmapIfPresent = [&](mlir::Type varType) {
+    using defMap = clause::Defaultmap;
+    auto exists = [&](defMap::VariableCategory varCat) {
+      return defaultMaps.find(varCat) != defaultMaps.end();
+    };
```
Nit: `llvm::is_contained()` should do what you need.

skatrak reviewed (commented):
Thank you Andrew, I've got some comments but the overall approach seems fine to me.

agozillon commented on flang/lib/Lower/OpenMP/ClauseProcessor.h:
```diff
@@ -32,6 +32,10 @@ namespace Fortran {
 namespace omp {
 
+// Container type for tracking user specified Defaultmaps for a target region
+using DefaultMapsTy = std::map<clause::Defaultmap::VariableCategory,
```
I think I'd prefer to keep it as a map, primarily as I think IndexMap would require me to either extend VariableCategory to have an operator overload for () that converts the stored enum to an integer as I don't think enum conversion to an integer is implicit (at least when used with indexed map) or a wrapper class that'd do the same but wrap VariableCategory, so we can avoid altering VariableCategory itself. Happy to do either of course, but seems like a little extra complexity! However, there's every chance I'm misunderstanding how IndexedMap works, it's my first time using it/tinkering with it (thank you for pointing it out, I tend to forget we have a lot of nice LLVM utilities) :-)

agozillon committed:
[Flang][OpenMP] Initial defaultmap implementation

This aims to implement most of the initial arguments for defaultmap aside from firstprivate and none, and some of the more recent OpenMP 6 additions which will come in subsequent updates.

agozillon commented:
Thank you very much for the review @skatrak it's greatly appreciated! I've updated the PR now with all of the above changes except the IndexedMap one currently which I've left a comment for my current reasoning! :-)

skatrak reviewed (approved):
Thanks again Andrew, LGTM. No need for another review by me after addressing these final nits.

agozillon commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -2231,6 +2232,146 @@ genSingleOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
 }
 
+static clause::Defaultmap::ImplicitBehavior
+getDefaultmapIfPresent(DefaultMapsTy &defaultMaps, mlir::Type varType) {
```
Couldn't add the const for either unfortunately, the index [] operator for map doesn't have a const qualified equivalent it seems (or one that utilises a const "this").

kparzysz replied:
std::map has "at" that works for "const this".  If the given key is not present in the map, it will abort.  The operator [] will insert the key with a default-initialized value, so the map cannot be const for it to work.

agozillon commented:
Thank you very much for the reviews @kparzysz @skatrak I'm going to go ahead and land this now, I addressed the remaining nit comments!
